plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -310,19 +311,22 @@ private static String buildSplitPath(String tableLocation, AddFileEntry addActio
     {
         // paths are relative to the table location and are RFC 2396 URIs
         // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-file-and-remove-file
-        URI uri = URI.create(addAction.getPath());
-        String path = uri.getPath();
+        String path = addAction.getPath();
+
+        URI uri = URI.create(path);
+
+        if (!uri.isAbsolute()) {
+            path = Locations.appendPath(tableLocation, uri.getPath());
+        }
 
         // org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem encodes the path as URL when opening files
         // https://issues.apache.org/jira/browse/HADOOP-18580
-        if (tableLocation.startsWith("abfs://") || tableLocation.startsWith("abfss://")) {
+        if (path.startsWith("abfs://") || path.startsWith("abfss://")) {
             // Replace '+' with '%2B' beforehand. Otherwise, the character becomes a space ' ' by URL decode.
             path = URLDecoder.decode(path.replace("+", "%2B"), UTF_8);
```

Comment:
@findepi indeed, this was breaking the old behaviour
```
trino> select * from delta.dev_30.base_table_test_vin_per_path;

Query 20230424_152802_00002_3bt3b, FAILED, 1 node
Splits: 2 total, 0 done (0.00%)
0.88 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20230424_152802_00002_3bt3b failed: Error opening Hive split abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%/base_table_test_vin+/part-00000-eeb717bb-9056-4959-bc79-a642879d8977-c000.snappy.parquet (offset=0, length=654): HEAD https://storagedfs.core.windows.net/container/prod-data/dev_30.db/%2Bab/a%25/base_table_test_vin%2B/part-00000-eeb717bb-9056-4959-bc79-a642879d8977-c000.snappy.parquet?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
```

changed it to 
```
if (path.startsWith("abfs://") || path.startsWith("abfss://")) {
            // Replace '+' with '%2B' beforehand. Otherwise, the character becomes a space ' ' by URL decode.
            path = URLDecoder.decode(uri.getPath().replace("+", "%2B"), UTF_8);
        }

        // paths are relative to the table
        if (!uri.isAbsolute()) {
            path = appendPath(tableLocation, path);
        }
```

post the change, the old behaviour is restored
```
trino> select * from delta.dev_30.base_table_test_vin_per_path;
 a | b 
---+---
 a | 3 
 b | 4 
(2 rows)
```

Reply:
@findepi @ebyhr there's a bug with existing code as well
```
trino> create table delta_prod.dev_30.base_table_test_vin_per_par_trino (a int, b varchar) WITH (partitioned_by = ARRAY['b']);
CREATE TABLE
trino> 
trino> 
trino> 
trino> insert into delta_prod.dev_30.base_table_test_vin_per_par_trino values (1, 'a%25'), (2, 'a%26');
INSERT: 2 rows

trino> 
trino> 
trino> 
trino> show create table delta_prod.dev_30.base_table_test_vin_per_par_trino;
                                                                              Create Table                                                                              
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 CREATE TABLE delta_prod.dev_30.base_table_test_vin_per_par_trino (                                                                                                     
    a integer,                                                                                                                                                          
    b varchar                                                                                                                                                           
 )                                                                                                                                                                      
 WITH (                                                                                                                                                                 
    location = 'abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90', 
    partitioned_by = ARRAY['b']                                                                                                                                         
 )                                                                                                                                                                      
(1 row)

trino> 
trino> 
trino> 
trino> 
trino> 
trino> 
trino> select * from delta_prod.dev_30.base_table_test_vin_per_par_trino;

Query 20230424_174752_01486_ti4fr failed: Error opening Hive split abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90/b=a%26/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d (offset=0, length=187): HEAD https://storage.dfs.core.windows.net/container/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90/b%3Da%2526/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
io.trino.spi.TrinoException: Error opening Hive split abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90/b=a%26/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d (offset=0, length=187): HEAD https://storage.dfs.core.windows.net/container/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90/b%3Da%2526/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createPageSource(ParquetPageSourceFactory.java:312)
	at io.trino.plugin.deltalake.DeltaLakePageSourceProvider.createPageSource(DeltaLakePageSourceProvider.java:203)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:49)
	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:298)
	at io.trino.operator.Driver.processInternal(Driver.java:402)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:305)
	at io.trino.operator.Driver.tryWithLock(Driver.java:701)
	at io.trino.operator.Driver.process(Driver.java:297)
	at io.trino.operator.Driver.processForDuration(Driver.java:268)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:845)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:537)
	at io.trino.$gen.Trino_412____20230423_061056_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.FileNotFoundException: HEAD https://storage.dfs.core.windows.net/container/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90/b%3Da%2526/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.checkException(AzureBlobFileSystem.java:926)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.open(AzureBlobFileSystem.java:177)
	at io.trino.hdfs.TrinoFileSystemCache$FileSystemWrapper.open(TrinoFileSystemCache.java:393)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:906)
	at io.trino.filesystem.hdfs.HdfsInputFile.lambda$openFile$1(HdfsInputFile.java:108)
	at io.trino.hdfs.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:25)
	at io.trino.hdfs.HdfsEnvironment.doAs(HdfsEnvironment.java:93)
	at io.trino.filesystem.hdfs.HdfsInputFile.openFile(HdfsInputFile.java:108)
	at io.trino.filesystem.hdfs.HdfsInputFile.newInput(HdfsInputFile.java:57)
	at io.trino.plugin.hive.parquet.TrinoParquetDataSource.<init>(TrinoParquetDataSource.java:39)
	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createPageSource(ParquetPageSourceFactory.java:227)
	... 16 more
Caused by: org.apache.hadoop.fs.azurebfs.contracts.exceptions.AbfsRestOperationException: HEAD https://storage.dfs.core.windows.net/container/prod-data/dev_30.db/base_table_test_vin_per_par_trino-3ba1dacbe24e45d6977fefbd1be52a90/b%3Da%2526/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d?timeout=90
StatusCode=404
StatusDescription=The specified path does not exist.
ErrorCode=
ErrorMessage=
	at org.apache.hadoop.fs.azurebfs.services.AbfsRestOperation.execute(AbfsRestOperation.java:134)
	at org.apache.hadoop.fs.azurebfs.services.AbfsClient.getPathProperties(AbfsClient.java:352)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystemStore.openFileForRead(AzureBlobFileSystemStore.java:349)
	at org.apache.hadoop.fs.azurebfs.AzureBlobFileSystem.open(AzureBlobFileSystem.java:174)
	... 25 more
```

Delta Log
```
{"commitInfo":{"version":1,"timestamp":1682358378856,"userId":"vinaykl","userName":"vinaykl","operation":"WRITE","operationParameters":{"queryId":"20230424_174618_01484_ti4fr"},"clusterId":"trino-412-trinomaster-interactive-deployment-6ccb9c6c85-cksfs","readVersion":0,"isolationLevel":"WriteSerializable","isBlindAppend":true}}
{"add":{"path":"b=a%252525/20230424_174618_01484_ti4fr-bc30e934-2795-47c0-b505-724df36b2b13","partitionValues":{"b":"a%25"},"size":187,"modificationTime":1682358378684,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":1},\"maxValues\":{\"a\":1},\"nullCount\":{\"a\":0}}","tags":{}}}
{"add":{"path":"b=a%252526/20230424_174618_01484_ti4fr-0e4bc0c6-358c-43be-9c59-472a9c13217d","partitionValues":{"b":"a%26"},"size":187,"modificationTime":1682358378684,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":2},\"maxValues\":{\"a\":2},\"nullCount\":{\"a\":0}}","tags":{}}}
```

Reply:
This is a known issue  https://github.com/trinodb/trino/issues/17012

cc @pajaks @dain

Reply:
```
private static String buildSplitPath(String tableLocation, AddFileEntry addAction)
    {
        // decoding is implicit within URI class, done lazily on get methods
        URI uri = URI.create(addAction.getPath());

        // paths are relative to the table
        if (!uri.isAbsolute()) {
            return appendPath(tableLocation, uri.getPath());;
        }
        else {
            return uri.getScheme() + ":" + uri.getSchemeSpecificPart();
        }
    }
```
@findinpath, the above code will covers the `table location containing special chars, partitions having special chars & shallow cloned from partitioned base table which references absolutePath with special chars`, This also potentially fixes https://github.com/trinodb/trino/issues/17012.

* Table location need not be `decoded` as it's not encoded for either `managed or unmanaged` Delta tables
* Only paths needs to be decoded, this is implicit when an URI is created for the and referenced subsequently
```
scala> spark.sql("create table dev_30.unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars (a int, b string) USING DELTA PARTITIONED BY (b) LOCATION 'abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%25/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/'")

scala> spark.sql("describe extended dev_30.unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars").show(1000, false)
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+
|col_name                    |data_type                                                                                                                                                                   |comment|
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+
|a                           |int                                                                                                                                                                         |       |
|b                           |string                                                                                                                                                                      |       |
|                            |                                                                                                                                                                            |       |
|# Partitioning              |                                                                                                                                                                            |       |
|Part 0                      |b                                                                                                                                                                           |       |
|                            |                                                                                                                                                                            |       |
|# Detailed Table Information|                                                                                                                                                                            |       |
|Name                        |dev_30.unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars                                                                                      |       |
|Location                    |abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%25/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+|       |
|Provider                    |delta                                                                                                                                                                       |       |
|Owner                       |myntra                                                                                                                                                                      |       |
|External                    |true                                                                                                                                                                        |       |
|Table Properties            |[delta.minReaderVersion=1,delta.minWriterVersion=2]                                                                                                                         |       |
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+

scala> spark.sql("select * from dev_30.unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars").show()
+----+-------------+
|   a|            b|
+----+-------------+
| 111|         a%25|
|  22|       a%2526|
|  11|       a%2525|
|   2|            b|
|   1|            a|
|1122|with?question|
| 222|         a%26|
+----+-------------+

--------------------------------- TRINO ----------------------------------

trino> select * from delta.dev_30.unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars;
  a   |       b       
------+---------------
  222 | a%26          
   22 | a%2526        
   11 | a%2525        
    1 | a             
  111 | a%25          
    2 | b             
 1122 | with?question
```

* Shallow Cloned table's delta_log JSON contains ` absolute paths` which are already encoded
```
scala> spark.sql("select * from dev_30.unmanaged_shallow_clone_delta_table_with_special_chars_with_partitioned_on_special_chars").show()
+----+-------------+
|   a|            b|
+----+-------------+
| 111|         a%25|
|  22|       a%2526|
|  11|       a%2525|
|   2|            b|
|   1|            a|
|1122|with?question|
| 222|         a%26|
+----+-------------+

--------------------- TRINO ---------------------------
trino> select * from delta.dev_30.unmanaged_shallow_clone_delta_table_with_special_chars_with_partitioned_on_special_chars;
  a   |       b       
------+---------------
    1 | a             
 1122 | with?question 
   11 | a%2525        
  111 | a%25          
  222 | a%26          
    2 | b             
   22 | a%2526        
(7 rows)

---------------------------------------------------------- delta_log entries of shallow cloned table -------------------

{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=a/part-00000-6416449f-93d6-4edb-907c-986db7784f6f.c000.snappy.parquet","partitionValues":{"b":"a"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":1},\"maxValues\":{\"a\":1},\"nullCount\":{\"a\":0}}"}}
{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=b/part-00001-de1e503e-7c15-4c3e-b5b7-f3f7adc99ac4.c000.snappy.parquet","partitionValues":{"b":"b"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":2},\"maxValues\":{\"a\":2},\"nullCount\":{\"a\":0}}"}}
{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=a%25252526/part-00006-bbc89c7a-6bbc-47b6-9dd0-08fc6a9ac460.c000.snappy.parquet","partitionValues":{"b":"a%2526"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":22},\"maxValues\":{\"a\":22},\"nullCount\":{\"a\":0}}"}}
{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=a%252526/part-00003-e753ca8f-e8f0-4ac6-8583-91cd8253117d.c000.snappy.parquet","partitionValues":{"b":"a%26"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":222},\"maxValues\":{\"a\":222},\"nullCount\":{\"a\":0}}"}}
{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=a%252525/part-00002-e2e9e420-94f0-4333-9a53-3dbdd10c4770.c000.snappy.parquet","partitionValues":{"b":"a%25"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":111},\"maxValues\":{\"a\":111},\"nullCount\":{\"a\":0}}"}}
{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=with%253Fquestion/part-00004-f80e913a-95c4-4fc2-abcb-77a56b9d22a7.c000.snappy.parquet","partitionValues":{"b":"with?question"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":1122},\"maxValues\":{\"a\":1122},\"nullCount\":{\"a\":0}}"}}
{"add":{"path":"abfs://container@storage.dfs.core.windows.net/prod-data/dev_30.db/+ab/a%2525/unmanaged_base_delta_table_with_special_chars_with_partitioned_on_special_chars+/b=a%25252525/part-00005-74f27b9f-9719-4a62-b24b-dea51729e577.c000.snappy.parquet","partitionValues":{"b":"a%2525"},"size":449,"modificationTime":1682372178000,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"a\":11},\"maxValues\":{\"a\":11},\"nullCount\":{\"a\":0}}"}}
```